# EJERCICIOS SIMPLES CON NUMPY 

In [1]:
#comenzamos por importar las librerias:
import numpy as np

1. Tenemos un array de una dimensión con la función range y que contenga 10. Queremos realizar los siguientes ejercicios

In [8]:
x = np.arange(0,10)
x

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

a) Mostrar los 5 primeros elementos

In [ ]:
x[0:6] #pongo 0 para establecer un rango, aunque también puedo usar :6

array([0, 1, 2, 3, 4, 5])

In [10]:
x[:6]

array([0, 1, 2, 3, 4, 5])

b) Mostrar elementos después del índice 5

In [11]:
x[5:]

array([5, 6, 7, 8, 9])

c) Cada dos elementos, comenzando en el índice 1

In [ ]:
x[1::2] #le opngo :: para dercirle que vaya de dos en dos

array([1, 3, 5, 7, 9])

d) Forma de invertir una matriz

In [ ]:
x[::-1] #es un slicing o rebanada


array([9, 8, 7, 6, 5, 4, 3, 2, 1, 0])

e) Mostrar un elemento si y uno no

In [14]:
x[::2]

array([0, 2, 4, 6, 8])

f) Mostrar matriz intermedia que vaya desde la posición 4 hasta la 7

In [15]:
x[4:8]

array([4, 5, 6, 7])

g) Mostrar la matriz invertida cada dos elementos desde el índice 5

In [16]:
x[5::-1]

array([5, 4, 3, 2, 1, 0])

2. Tenemos un array de dos dimensiones y queremos que nos muestre:

In [17]:

x2d=np.random.randint(10, size=(3,4))

In [18]:
x2d

array([[3, 8, 0, 0],
       [1, 7, 7, 1],
       [8, 9, 4, 9]], dtype=int32)

a) dos filas, tres columnas

In [19]:
x2d[1:3, 0:3]

array([[1, 7, 7],
       [8, 9, 4]], dtype=int32)

b) Todas las filas, cada dos columnas

In [22]:
x2d[:,::2]

array([[3, 0],
       [1, 7],
       [8, 4]], dtype=int32)

3.Filtrar números impares: Dado un array 2D, escribe un programa que filtre y devuelva todos los números impares.

In [24]:
# Crear un array 2D
array = np.array([[10,15,20],[25,30,35]])
array

array([[10, 15, 20],
       [25, 30, 35]])

In [26]:
# Filtrar números impares:
impares = array[array %2 !=0] #decimos que si se divide entre 2, el resto sea diferente a 0
impares


array([15, 25, 35])

4.Filtrar números mayores a la media del array: Crea un programa que encuentre y devuelva todos los elementos mayores al promedio del array.

In [27]:
array = np.array([50,50,60,80,90,9090,50])
array

array([  50,   50,   60,   80,   90, 9090,   50])

In [ ]:
# Calcular la media
media = array.mean()
media

np.float64(1352.857142857143)

In [ ]:
# Filtrar elementos mayores a la media. Construimos una máscara y le pasamos la variable media

mayores = array[array > media]
mayores


array([9090])

In [31]:
#tamabién lo puedo hacer con booleanos
mayores1 =  [array > media]
mayores1

[array([False, False, False, False, False,  True, False])]

5.Suma de arrays y escalar: Dado un array de dimensiones 3×3 suma otro array de las mismas dimensiones y multiplica el resultado por 2.

In [36]:
a = np.array([[1,7,9],[8,10,25],[1,7,9]])
b = np.array([[1,7,6],[8,10,45],[1,7,91]])

In [ ]:
# Sumar los arrays y multiplicar por 2






6.Elevar un array al cuadrado y sumar filas: Crea un programa que eleve al cuadrado todos los elementos de un array y luego calcule la suma de cada fila.

In [ ]:
# Crear un array



In [ ]:
# Elevar al cuadrado





In [ ]:
#suma de filas


In [ ]:
# Sumar total



7.Sistemas de Ecuaciones:
Resolver un sistema de ecuaciones lineales
Dado el siguiente sistema de ecuaciones:


3x+2y=12

x−y=1

In [38]:
from numpy.linalg import solve
# Matriz de coeficientes
a=np.array([[3,2],[1,-1]])

# Vector de términos independientes
b=np.array([12,1])


In [39]:
# Resolver el sistema
solucion = solve (a,b)
solucion

array([2.8, 1.8])

8.Resolver un sistema de tres ecuaciones
Resuelve el siguiente sistema de ecuaciones:


x+y+z=6

2x+y−z=3

x−y+2z=7

In [42]:
# Matriz de coeficientes
from numpy.linalg import solve
# Matriz de coeficientes
a = np.array([[1,1,1,],[2,1,-1],[1,-1,2]])
b = np.array([6,3,7])

# Vector de términos independientes
soluion=solve(a,b)
solucion



array([2.8, 1.8])

In [ ]:
# Resolver el sistema


# -------------------------------------------------------------------------------

# Ejemplo: ¿Cuál es la altura promedio de los presidentes de Estados Unidos?
Los array disponibles en NumPy pueden resultar extremadamente útiles para resumir un conjunto de valores.
Como ejemplo sencillo, consideremos las alturas de todos los presidentes de Estados Unidos.
Estos datos están disponibles en el archivo *president_heights.csv*, que es una lista simple de etiquetas y valores separados por comas:

In [43]:
# Importamos las librerías necesarias
import pandas as pd
import numpy as np


In [45]:
# Creamos una variable donde leemos el archivo en este caso formato csv
df = pd.read_csv('president_heights.csv')
df.sample(5)

order               name  height(cm)
34     37      Richard Nixon         182
0       1  George Washington         189
25     28     Woodrow Wilson         180
41     44       Barack Obama         185
32     35    John F. Kennedy         183

In [47]:
# Tomamos de la tabla la columna de height(cm) y la convertimos en un array
altura = np.array(df['height(cm)'])
altura

array([189, 170, 189, 163, 183, 171, 185, 168, 173, 183, 173, 173, 175,
       178, 183, 193, 178, 173, 174, 183, 183, 168, 170, 178, 182, 180,
       183, 178, 182, 188, 175, 179, 183, 193, 182, 183, 177, 185, 188,
       188, 182, 185])

Ahora que tenemos esta matriz de datos, podemos calcular una variedad de estadísticas resumidas:

In [50]:
#Altura media
media = round(altura.mean(),2)
media

np.float64(179.74)

In [51]:
max = round(altura.max(),2)
max

np.int64(193)

In [ ]:
min=round(altura.min(),2)
min


In [54]:
std=round(altura.std(),2)
std


np.float64(6.93)

Otros cálculos que podermos realizar:

In [56]:
#percentil 25
per_25 = np.percentile(altura,25)
per_25


np.float64(174.25)

In [57]:
per_75 = np.percentile(altura,75)
per_75


np.float64(183.0)

In [58]:
mediana = np.median(altura)
mediana

np.float64(182.0)


Vemos que la altura media, redondeando, de los presidentes estadounidenses es de 182 cm.

Por supuesto, a veces es más útil ver una representación visual de estos datos, lo que podemos lograr usando herramientas en Matplotlib. 

Por ejemplo, este código genera el siguiente gráfico:

In [ ]:
# Importamos la librería matplotlib utilizada para la generación de gráficos




SyntaxError: invalid syntax (477927217.py, line 2)

¿Qué tan selecto es el club de presidentes que superan los 185 cm de altura?

Calcula cuántos presidentes pertenecen a este "club" y cuál es su altura media.

In [ ]:
masaltos = np.array(df[df['height(cm)']>185]) #Opcion 1
masaltos

array([[1, 'George Washington', 189],
       [3, 'Thomas Jefferson', 189],
       [16, 'Abraham Lincoln', 193],
       [32, 'Franklin D. Roosevelt', 188],
       [36, 'Lyndon B. Johnson', 193],
       [41, 'George H. W. Bush', 188],
       [42, 'Bill Clinton', 188]], dtype=object)

In [ ]:
masaltos2 = df['height(cm)'].values  #opcion 2 trayendo solo los valores
masaltos2 = masaltos2[masaltos2>185]
masaltos2

array([189, 189, 193, 188, 193, 188, 188])

In [71]:
masaltos2.mean()

np.float64(189.71428571428572)

Se dice que los primeros presidentes eran más altos.
Calcula la altura media de los primeros 10 presidentes y compárala con los últimos 10.

In [73]:
todas_alturas = df['height(cm)'].values
todas_alturas


array([189, 170, 189, 163, 183, 171, 185, 168, 173, 183, 173, 173, 175,
       178, 183, 193, 178, 173, 174, 183, 183, 168, 170, 178, 182, 180,
       183, 178, 182, 188, 175, 179, 183, 193, 182, 183, 177, 185, 188,
       188, 182, 185])

In [74]:
primeros_10 = todas_alturas[:10]
primeros_10.mean()

np.float64(177.4)

In [75]:
ultimos_10 = todas_alturas[-10:]
ultimos_10.mean()

np.float64(184.6)

¿Quién está más cerca de la media?

Usando NumPy, encuentra al presidente cuya altura está más cerca de la altura media de todos.

Ahora usando el Dataset de votos debemo unificar la informacion de ambos DF

¿Quién saca más partido de su altura?

Calcula cuántos millones de votos recibe cada presidente por centímetro de altura.
¿Quién es el más "eficiente"? ¿Y el menos?

Calcula la correlación entre altura y votos usando NumPy. ¿Existe alguna relación?

¿Qué tan desigual es la distribución de votos?

Calcula el coeficiente de variación (CV) de los votos

El top 5 presidencial

¿Quiénes son los 5 presidentes con más votos totales?
¿Cuáles son sus alturas? ¿Hay alguna tendencia?